File: src/main/java/me/chell/samsara/impl/gui/buttons/Slider.java
package me.chell.samsara.impl.gui.buttons;

import com.google.common.collect.Lists;
import me.chell.samsara.api.gui.GuiTheme;
import me.chell.samsara.api.value.Value;
import me.chell.samsara.impl.gui.ValueButton;
import net.minecraft.client.util.InputUtil;
import net.minecraft.util.math.MathHelper;
import org.apache.commons.lang3.StringUtils;
import org.lwjgl.glfw.GLFW;

import java.awt.*;
import java.util.List;

public class Slider<T extends Number> extends ValueButton<T> {
    private int sliderWidth = 0;
    private boolean grabbed = false;

    private int ticks;
    private boolean editing = false;
    private String oldText, editText;

    public void setSliderWidth() {
        double min = value.getSliderMin().doubleValue();
        double max = value.getSliderMax().doubleValue();
        double val = value.getValue().doubleValue() - min;
        double diff = min >= 0 ? max - min : min - max;
        if(val == 0) sliderWidth = 0;
        double d = val / diff;
        if(d < 0) d *= -1;
        sliderWidth = (int) ((width-2) * MathHelper.clamp(d, 0D, 1D));
    }

    public Slider(Value<T> value, int x, int y) {
        super(x, y, GuiTheme.width, 13, value);
        editText = ""+value.getValue();

        setSliderWidth();
    }

    @SuppressWarnings("unchecked")
    @Override
    public void draw(int mouseX, int mouseY) {
        if(value.getName().equals("Rotation")) {
            double min = value.getSliderMin().doubleValue();
            double max = value.getSliderMax().doubleValue();
            double val = value.getValue().doubleValue() - min;
            double diff = min >= 0 ? max - min : min - max;
            if(val == 0) sliderWidth = 0;
            double d = val / diff;
            if(d < 0) d *= -1;
            sliderWidth = (int) ((width-2) * MathHelper.clamp(d, 0D, 1D));
        }

        if(grabbed) {
            int mousePos = MathHelper.clamp(mouseX - (x+2), 0, width-2);

            double min = value.getSliderMin().doubleValue();
            double max = value.getSliderMax().doubleValue();
            double diff = max - min;

            double v = mousePos / (width-2D);
            v *= diff;
            v += min;

            if(value.getSliderMin() instanceof Integer) {
                value.setValue((T) (Number) (int) v);
            } else
                value.setValue((T) (Number) (Math.round(v * 100d) / 100d));

            setSliderWidth();
        }

        drawThemedRectPrimary(x+2, y, sliderWidth, height-1);
        drawThemedRectSecondary(x+2+sliderWidth, y, width-2-sliderWidth, height-1);
        drawThemedRectTertiary(x, y, 2, height);
        drawThemedRectTertiary(x+2, y+height-1, width-2, 1);

        String displayText;
        if(editing) {
            displayText = editText;
            if(ticks / 6 % 2 == 0) displayText += "_";
        } else {
            displayText = ""+value.getValue();
        }

        drawThemedString(value.getDisplayName(), x + 4, getStringCenterY(y, height-1));
        int rightX = x + width - 2;
        if(displayText.endsWith("_")) rightX += getFontRenderer().getWidth("_");
        drawThemedStringRight(displayText, rightX, getStringCenterY(y, height-1));
    }

    @Override
    public boolean mouseClicked(int mouseX, int mouseY, int mouseButton) {
        if(mouseX >= x+2 && mouseX <= x + width && mouseY >= y && mouseY <= y + height-1) {
            if(mouseButton == 0) {
                grabbed = true;
                return true;
            } else if(mouseButton == 2) {
                if(!editing) {
                    ticks = 0;
                    editing = true;
                    editText = oldText = ""+value.getValue();
                }
                return true;
            }
        }
        return super.mouseClicked(mouseX, mouseY, mouseButton);
    }

    @Override
    public void mouseReleased(int mouseX, int mouseY, int state) {
        if(state == 0) {
            grabbed = false;
        }
    }

    private final List<Character> numbers = Lists.newArrayList('0', '1', '2', '3', '4', '5', '6', '7', '8', '9', '.', '-');

    @Override
    public boolean keyTyped(char typedChar, int keyCode) {
        if(editing) {
            switch (keyCode) {
                case GLFW.GLFW_KEY_BACKSPACE:
                    editText = StringUtils.chop(editText);
                    return true;
                case GLFW.GLFW_KEY_ENTER:
                    parseValue(editText);
                    editing = false;
                    return true;
                case GLFW.GLFW_KEY_ESCAPE:
                    editing = false;
                    editText = oldText;
                    return true;
                case GLFW.GLFW_KEY_V:
                    if (InputUtil.isKeyPressed(getMc().getWindow().getHandle(), GLFW.GLFW_KEY_LEFT_CONTROL)) {
                        editText += Toolkit.getDefaultToolkit().getSystemClipboard().getContents(null);
                    } else {
                        editText += typedChar;
                    }
                    return true;
                default:
                    if(numbers.contains(typedChar)) {
                        editText += typedChar;
                        return true;
                    }
            }
        }
        return super.keyTyped(typedChar, keyCode);
    }

    @SuppressWarnings("unchecked")
    private void parseValue(String text) {
        if (value.getSliderMin() instanceof Integer) {
            try {
                value.setValue((T) (Integer) Integer.parseInt(text));
                setSliderWidth();
            } catch (NumberFormatException ignored) {}
        }
        else if (value.getSliderMin() instanceof Float) {
            try {
                value.setValue((T) (Float) Float.parseFloat(text));
                setSliderWidth();
            } catch (NumberFormatException ignored) {}
        }
        else if(value.getSliderMin() instanceof Double) {
            try {
                value.setValue((T) (Double) Double.parseDouble(text));
                setSliderWidth();
            } catch (NumberFormatException ignored) {}
        }
    }

    @Override
    public void onGuiClosed() {
        grabbed = false;
        editing = false;
        editText = oldText;
    }

    @Override
    public void updateScreen() {
        ticks++;
    }
}

File: src/main/kotlin/me/chell/samsara/impl/gui/click/buttons/value/Slider.kt
package me.chell.samsara.impl.gui.click.buttons.value

import me.chell.samsara.api.gui.Rectangle
import me.chell.samsara.api.value.Value
import me.chell.samsara.impl.gui.click.Button
import me.chell.samsara.impl.gui.click.Window
import me.chell.samsara.impl.gui.click.buttons.ValueButton
import net.minecraft.client.util.math.MatrixStack
import net.minecraft.util.math.MathHelper
import org.lwjgl.glfw.GLFW
import java.lang.NumberFormatException
import kotlin.math.roundToInt

class Slider(val v: Value<Number>, override var x: Double, override var y: Double): ValueButton<Number>(v, x, y) {

    private var grabbed = false
    private var sliderWidth = 0.0

    private var listening = false
    private var input = ""
    private var tickCounter = 0

    override fun render(matrices: MatrixStack, mouseX: Double, mouseY: Double, tickDelta: Float) {
        val rect = Rectangle(x, y, Window.width, height).subtract(border)
        rect.x += Window.padding.left + Button.border.left
        rect.width -= Window.padding.left + Window.padding.right + Button.border.left + Button.border.right

        if (grabbed) {
            val mousePos = MathHelper.clamp(mouseX - rect.x, 0.0, rect.width)

            val percent = mousePos / rect.width

            val v = (value.max!!.toDouble() - value.min!!.toDouble()) * percent + value.min.toDouble()

            setValue((v * 100.0).roundToInt() / 100.0)
        }

        //draw border
        drawBorder(matrices, border, rect, borderColor)

        // draw background
        rect.x += sliderWidth
        rect.width -= sliderWidth
        fill(matrices, rect, secondaryColor)

        // draw slider
        if (sliderWidth > 0.0) {
            val sliderRect = Rectangle(rect.x, rect.y, rect.width, rect.height)
            sliderRect.x -= sliderWidth
            sliderRect.width = sliderWidth

            fill(matrices, sliderRect, primaryColor)
        }

        // draw text
        rect.x -= sliderWidth
        rect.width += sliderWidth
        drawString(matrices, value.displayName, primaryText, rect)

        val w = fontRenderer.getWidth(value.displayName)
        rect.x += w
        rect.width -= w

        var text = if(listening) input else value.value.toString()
        if(listening && tickCounter % 10 > 3) text += '\u007C'

        drawString(matrices, text, secondaryText, rect)
    }

    override fun tick() {
        tickCounter++
    }

    override fun screenResized() {
        setSliderWidth()
        tickCounter = 0
    }

    override fun themeLoaded() {
        super.themeLoaded()
        if(isVisible())
            setSliderWidth()
    }

    override fun guiClosed() {
        listening = false
        grabbed = false
    }

    override fun mouseClicked(mouseX: Double, mouseY: Double, button: Int): Boolean {
        if(Rectangle(x, y, Window.width, height).isInBounds(mouseX, mouseY)) {
            when(button) {
                0 -> {
                    grabbed = true
                    return true
                }
                1 -> {
                    listening = true
                    input = value.value.toString()
                    return true
                }
            }
        }
        return false
    }

    override fun mouseReleased(mouseX: Double, mouseY: Double, button: Int): Boolean {
        if(grabbed) {
            grabbed = false
            return true
        }
        return false
    }

    private val whitelist = listOf('0', '1', '2', '3', '4', '5', '6', '7', '8', '9', '.', '-', 'e')

    override fun keyPressed(keyCode: Int, scanCode: Int, modifiers: Int): Boolean {
        if(!listening) return false

        return when(keyCode) {
            GLFW.GLFW_KEY_ENTER, GLFW.GLFW_KEY_KP_ENTER -> {
                try {
                    when(value.min) {
                        is Int -> setValue(input.toInt())
                        is Double -> setValue(input.toDouble())
                        is Float -> setValue(input.toFloat())
                    }
                } catch (ignored: NumberFormatException) {}

                listening = false
                true
            }
            GLFW.GLFW_KEY_BACKSPACE -> {
                input = input.dropLast(1)
                true
            }
            GLFW.GLFW_KEY_ESCAPE -> {
                listening = false
                true
            }
            else -> false
        }
    }

    override fun charTyped(char: Char, modifiers: Int): Boolean {
        if(listening && whitelist.contains(char)) {
            input += char
            return true
        }
        return false
    }

    private fun setSliderWidth() {
        val min = value.min!!.toDouble()
        val max = value.max!!.toDouble()
        val v = value.value.toDouble()
        val diff = max - min

        val percent = (v - min) / diff
        val width = Window.width - Window.padding.left - Window.padding.right - Button.border.left - Button.border.right - border.left - border.right

        sliderWidth = MathHelper.clamp(width * percent, 0.0, width)
    }

    private fun setValue(v: Number) {
        value.value = if(value.min is Int) v.toInt() else v
        setSliderWidth()
    }
}